ksqldb-engine/src/main/java/io/confluent/ksql/internal/UtilizationMetricsListener.java

```diff
@@ -15,52 +15,137 @@
   private final List<String> metrics;
-  private final Time time;
-  private long lastSampleTime;
+  private Instant lastSampleTime;
+  private final File baseDir;
 
-  public UtilizationMetricsListener() {
-    this.kafkaStreams = new ArrayList<>();
-    this.metrics = new LinkedList<>();
-    time = Time.SYSTEM;
-    lastSampleTime = time.milliseconds();
+  public UtilizationMetricsListener(final KsqlConfig config) {
+    this.kafkaStreams = new ConcurrentHashMap<>();
+    this.metrics = new ArrayList<>();
+    this.streamsDirectories = new ArrayList<>();
+    this.baseDir = new File(config.getKsqlStreamConfigProps()
+      .getOrDefault(
+        StreamsConfig.STATE_DIR_CONFIG,
+        StreamsConfig.configDef().defaultValues().get(StreamsConfig.STATE_DIR_CONFIG))
+      .toString());
+  }
+
+  // for testing
+  public UtilizationMetricsListener(
+      final ConcurrentHashMap<QueryId, KafkaStreams> kafkaStreams,
+      final List<File> streamsDirectories,
+      final File baseDir) {
+    this.kafkaStreams = kafkaStreams;
+    this.metrics = new ArrayList<>();
+    this.streamsDirectories = streamsDirectories;
+    this.baseDir = baseDir;
   }
 
   @Override
   public void onCreate(
       final ServiceContext serviceContext,
       final MetaStore metaStore,
       final QueryMetadata queryMetadata) {
-    kafkaStreams.add(queryMetadata.getKafkaStreams());
+    kafkaStreams.put(queryMetadata.getQueryId(), queryMetadata.getKafkaStreams());
+    streamsDirectories.add(new File(baseDir, queryMetadata.getQueryApplicationId()));
   }
 
   @Override
   public void onDeregister(final QueryMetadata query) {
-    final KafkaStreams streams = query.getKafkaStreams();
-    kafkaStreams.remove(streams);
+    kafkaStreams.remove(query.getQueryId());
+    streamsDirectories.remove(new File(baseDir, query.getQueryApplicationId()));
   }
 
   @Override
   public void run() {
     logger.info("Reporting Observability Metrics");
-    final Long currentTime = time.milliseconds();
-    // here is where we would report metrics
+    final Instant currentTime = Instant.now();
+    logger.info("Reporting CSU thread level metrics");
+    final List<MetricsReporter.DataPoint> dataPoints = new ArrayList<>();
+    nodeDiskUsage(dataPoints, currentTime);
+    taskDiskUsage(dataPoints, currentTime);
+    // here is where we would call report() for telemetry reporter
     lastSampleTime = currentTime;
   }
+
+  public void nodeDiskUsage(
+      final List<MetricsReporter.DataPoint> dataPoints,
+      final Instant sampleTime) {
+    for (File f : streamsDirectories) {
+      final double freeSpace = f.getFreeSpace();
+      final long totalSpace = f.getTotalSpace();
+      final double percFree = percentage(freeSpace, (double) totalSpace);
+      dataPoints.add(new MetricsReporter.DataPoint(sampleTime,"storage-usage", freeSpace));
+      dataPoints.add(new MetricsReporter.DataPoint(sampleTime,"storage-usage-perc", percFree));
```

Comment:
I found that looking at the raw free space number didn't give me a sense of how full my disk was, but maybe other people automatically know how much space there is total? For task level we don't have total file size so maybe we don't want it here either for consistency